### ISSUE TITLE
support serializing embedding of unexported struct types

### ISSUE THREAD

**Issue by edwingeng** [go-json-experiment/json#55]
```
type fooImpl struct{}

type Foo struct {
	fooImpl `json:"-"`
	x       int
}

func main() {
	var f Foo
	data, err := json.Marshal(&f)
	if err != nil {
		panic(err)
	}
	fmt.Println(string(data))
}
```

`json.Marshal` in above code does not return an error, which is not in consistent with the v2 design.

> A Go struct with only unexported fields cannot be serialized.

**Comment by dsnet** [comment]
This has a field explicitly marked as the `json` tag, so we treat this as a situation where the author of `Foo` deliberately wanted this to serialize to an empty object.

The main motivation for

> A Go struct with only unexported fields cannot be serialized.

is to avoid accidentally serializing a Go type without any indication that it was meant for JSON.

**Comment by edwingeng** [comment]
Actually, the author is forced to set the tag, because:

> A Go struct that embeds an unexported struct type cannot be serialized.

**Comment by dsnet** [comment]
Oh, I just saw that you're embedding an unexported type. You are running into this:

https://github.com/go-json-experiment/json/blob/ebd3a8989ca1eadb7a68e02a93448ecbbab5900c/diff_test.go#L1152-L1234

That has consistently been a can of worms in v1. The way it worked depended on bugs in the Go reflection API, which is not a good stance to preserve in v2, since it means it was always standing on shaky grounds.

**Comment by edwingeng** [comment]
Perhaps we need to replace the rule:

> A Go struct that embeds an unexported struct type cannot be serialized.

with some more specific rules:

- Unexported pointer struct-type embedding in a struct is always ignored silently.
- Unexported non-pointer struct-type embedding in a struct always needs a json tag to clarity its `Marshal/Unmarshal` behavior. If its json field name is `*`, it is treaded as it was in v1.
- Only one unexported non-pointer struct-type embedding in a struct can be tagged like `json:"*"`.
- If an unexported non-pointer struct-type embedding is tagged like `json:"foo"`, `Marshal` marshals it into a child node. `Unmarshal` unmarshals it from a child node.

Marshaling the following `Foo`:

```
type fooImpl struct {
	N int
}

type bar struct {
	N int
}

type Foo struct {
	fooImpl `json:"*"`
	bar     `json:"bar"`
	x       int
}
```

may output:

```
{
    "N": 0,
    "bar": {
        "N": 0
    }
}
```

**Comment by dsnet** [comment]
> Unexported pointer struct-type embedding in a struct is always ignored silently.

This is unsafe as users migrating from v1 to v2 will not be made aware that something that formerly worked in v1 no longer works in v2.

> Unexported non-pointer struct-type embedding in a struct

Unexported non-pointer struct embedding still doesn't quite work the same as exported embedded structs as you cannot roundtrip `Addr()` and `Elem()` calls. The v1 logic jumps through hacky hoops to partially emulate this behavior, but it still has edge cases, especially with regards to addressability, which is a significant property that v2 aims to consistently handle correctly everywhere.

Unexported embedding is just asking for trouble, and trying to carve out exceptions to make some things kind of work is a worse position to be in. The current position of v2 is at least simple, if even it does prevent some use-cases.

**Comment by edwingeng** [comment]
The impact of doing so may exceed expectation. Many golang developers use embedding as the cure for inheritance. This may break the convention.

**Comment by edwingeng** [comment]
I didn't notice the inline tag. It's a much better design than `json:"*"`.

**Comment by dsnet** [comment]
So I did more digging into the history of reflection and we're going to add back support for this.

- In Go 1.10 (via golang/go#24153), reflection on such fields became more difficult, which required some gross hacks in v1 "encoding/json" to somewhat emulate the behavior.
- In Go 1.15 (via golang/go#32772), the former restriction in reflect was reverted. The workaround hack still exists in v1 "encoding/json", but technically it could be removed now.

When we made the decision to forbid exported fields of embedded unexported types, we were motivated only by golang/go#24153 and were not aware that golang/go#32772 had occurred. Since the latter has finalized a decision to allow Go reflection access to such fields, we can more readily support these fields without fear that we'll be broken by the Go reflection again in the future.
